client.go

```diff
@@ -536,56 +537,142 @@ func (c *HelmClient) upgradeCRDs(ctx context.Context, chartInstance *chart.Chart
 	}
 
 	for _, crd := range chartInstance.CRDObjects() {
-		// use this ugly detour to parse the crdYaml to a CustomResourceDefinitions-Object because direct
-		// yaml-unmarshalling does not find the correct keys
-		jsonCRD, err := yaml.ToJSON(crd.File.Data)
-		if err != nil {
+		if err := c.upgradeCRD(ctx, k8sClient, crd); err != nil {
 			return err
 		}
+		log.Printf("CRD %s upgraded successfully for chart: %s", crd.Name, chartInstance.Metadata.Name)
+	}
+
+	return nil
+}
+
+func (c *HelmClient) upgradeCRD(ctx context.Context, k8sClient *clientset.Clientset, crd chart.CRD) error {
+	// use this ugly detour to parse the crdYaml to a CustomResourceDefinitions-Object because direct
+	// yaml-unmarshalling does not find the correct keys
+	jsonCRD, err := yaml.ToJSON(crd.File.Data)
+	if err != nil {
+		return err
+	}
+
+	var typeMeta metav1.TypeMeta
+	err = json.Unmarshal(jsonCRD, &typeMeta)
+	if err != nil {
+		return err
+	}
+
+	switch typeMeta.APIVersion {
+	default:
+		return fmt.Errorf("WARNING: failed to upgrade CRD %q: unsupported api-version %q", crd.Name, typeMeta.APIVersion)
 
-		var meta metaV1.TypeMeta
-		err = json.Unmarshal(jsonCRD, &meta)
+	case "apiextensions.k8s.io/v1beta1":
+		var crdObj v1beta1.CustomResourceDefinition
```

Comment:
Both branches of the switch/case are largely duplicate, but I guess they're difficult to abstract into a single function because they're working on different API types? Maybe extract both branches of the switch/case into separate functions, to at least break down this long function a bit?

Reply:
> Both branches of the switch/case are largely duplicate, but I guess they're difficult to abstract into a single function because they're working on different API types?

Exactly :+1: 
> Maybe extract both branches of the switch/case into separate functions, to at least break down this long function a bit?

I've updated the PR as suggested.